File: cpp/src/pivot.cpp
#include "simplex/pivot.hpp"
#include <cmath>
#include <vector>

std::optional<int> LargestCoefficient::entering_index(const Eigen::VectorXd& cr,
                                                        double tol) const
{
    int    best_j = -1;
    double best_v = tol;
    for (int j = 0; j < static_cast<int>(cr.size()); ++j) {
        if (cr[j] > best_v) {
            best_v = cr[j];
            best_j = j;
        }
    }
    if (best_j < 0) return std::nullopt;
    return best_j;
}

std::optional<int> Bland::entering_index(const Eigen::VectorXd& cr,
                                          double tol) const
{
    for (int j = 0; j < static_cast<int>(cr.size()); ++j) {
        if (cr[j] > tol) return j;
    }
    return std::nullopt;
}

std::optional<int> leaving_index(const Eigen::VectorXd& xb,
                                  const Eigen::VectorXd& d,
                                  const Eigen::MatrixXd& BinvN,
                                  double tol)
{
    int m = static_cast<int>(d.size());

    std::vector<int> candidates;
    candidates.reserve(m);
    for (int i = 0; i < m; ++i) {
        if (d[i] > tol) candidates.push_back(i);
    }
    if (candidates.empty()) return std::nullopt;

    int best = candidates[0];
    for (std::size_t k = 1; k < candidates.size(); ++k) {
        int i = candidates[k];
        double ratio_i    = xb[i]    / d[i];
        double ratio_best = xb[best] / d[best];

        if (ratio_i < ratio_best - tol) {
            best = i;
        } else if (std::abs(ratio_i - ratio_best) <= tol) {
            // Lexicographic tie-break: compare scaled rows of BinvN exactly,
            // mirroring Julia's Tuple(row_i) < Tuple(row_best).
            Eigen::VectorXd row_i    = BinvN.row(i).transpose()   / d[i];
            Eigen::VectorXd row_best = BinvN.row(best).transpose() / d[best];
            for (int c = 0; c < static_cast<int>(row_i.size()); ++c) {
                if (row_i[c] < row_best[c]) { best = i; break; }
                if (row_i[c] > row_best[c]) {           break; }
            }
        }
    }
    return best;
}

File: rust/src/pivot.rs
use nalgebra::DMatrix;

pub trait PivotRule {
    fn entering_index(&self, cr: &[f64], tol: f64) -> Option<usize>;
}

/// Pick the non-basic variable with the largest (most positive) reduced cost.
pub struct LargestCoefficient;

/// Bland's rule: always pick the lowest-index positive reduced cost.
pub struct Bland;

impl PivotRule for LargestCoefficient {
    fn entering_index(&self, cr: &[f64], tol: f64) -> Option<usize> {
        let (j, &v) = cr.iter().enumerate().max_by(|(_, a), (_, b)| a.partial_cmp(b).unwrap())?;
        if v > tol { Some(j) } else { None }
    }
}

impl PivotRule for Bland {
    fn entering_index(&self, cr: &[f64], tol: f64) -> Option<usize> {
        cr.iter().position(|&v| v > tol)
    }
}

/// Lexicographic ratio test — selects the leaving variable to prevent cycling.
///
/// Returns the row position (into `basic`) of the leaving variable, or `None`
/// if every component of `d` is non-positive (problem is unbounded).
pub fn leaving_index(xb: &[f64], d: &[f64], binv_n: &DMatrix<f64>, tol: f64) -> Option<usize> {
    let candidates: Vec<usize> = (0..d.len()).filter(|&i| d[i] > tol).collect();
    if candidates.is_empty() {
        return None;
    }

    let mut best = candidates[0];
    for &i in &candidates[1..] {
        let ratio_i = xb[i] / d[i];
        let ratio_best = xb[best] / d[best];

        if ratio_i < ratio_best - tol {
            best = i;
        } else if (ratio_i - ratio_best).abs() <= tol {
            // Lexicographic tie-break on scaled rows of BinvN
            let ncols = binv_n.ncols();
            let row_i: Vec<f64> = (0..ncols).map(|k| binv_n[(i, k)] / d[i]).collect();
            let row_best: Vec<f64> = (0..ncols).map(|k| binv_n[(best, k)] / d[best]).collect();
            if row_i < row_best {
                best = i;
            }
        }
    }

    Some(best)
}
